text/0673-deprecate-tryinvoke.md

```diff
@@ -0,0 +1,93 @@
+
+## Motivation
+
+In most cases, function arguments should not be optional, but in the rare occasion that it is optional, the Javascript language has [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) so we can deprecate the usage of `tryInvoke`.
+
+## Transition Path
+
+Ember will start logging deprecation messages for `tryInvoke` usage. Deprecation text: `Using tryInvoke has been deprecated. Instead, consider using native Javascript optional chaining.`
+
+We can codemod our current usage of `tryInvoke` with the equivalent behaviour using plain JavaScript. The migration guide will cover this example:
+
+Before:
+
+```js
+import { tryInvoke } from '@ember/utils';
+ 
+foo() {
+ tryInvoke(this.args, 'bar', ['baz']);
+}
+```
+
+After:
+
+```js
+foo() {
+ this.args.bar?.('baz');
+}
+```
+
+#### Using Optional Chaining Operator
+
+The optional chaining operator `?.` permits reading the value of a property located deep within a chain of connected objects without having to expressly validate that each reference in the chain is valid. The `?.` operator functions similarly to the `.` chaining operator, except that instead of causing an error if a reference is nullish (`null` or `undefined`), the expression short-circuits with a return value of `undefined`. When used with function calls, it returns `undefined` if the given function does not exist:
+
+```js
+const adventurer = {
+  name: 'Alice',
+  cat: {
+    name: 'Dinah'
+  }
+};
+
+const dogName = adventurer.dog?.name;
+console.log(dogName);
+// expected output: undefined
+
+console.log(adventurer.someNonExistentMethod?.());
+// expected output: undefined
+```
+
+Tooling Support:
+
+- [Babel](https://babeljs.io/) already supports the [optional chaining operator](https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining) so we can use that for future use-cases.
+
+- [ember-cli-babel](https://www.npmjs.com/package/ember-cli-babel) all recent versions support optional chaining operator.
+
+- [TypeScript](https://github.com/microsoft/TypeScript), similarly, as of [version 3.7](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) also supports the operator so we will not be breaking that flow either.
+
+## How We Teach This
+
+Add the transition path to the [Ember Deprecation Guide](https://deprecations.emberjs.com/). Deprecation text: `Using tryInvoke has been deprecated. Instead, consider using native Javascript optional chaining.`
+
+The references to `tryInvoke` will need to be removed from the [API docs](https://api.emberjs.com/ember/release/functions/@ember%2Futils/tryInvoke). 
+
+There are no changes needed for the [Ember Guides](https://guides.emberjs.com/release/) since we do not use it anywhere.
+
+## Drawbacks
+
+This change will cause some deprecation noise but could be mitigated with a codemod.
+
+## Alternatives
```

Comment:
We should maybe add the alternative to use Ember assert here similar to typeof Function

Reply:
But also this alternatives section I think has traditionally been for "alternative to deprecating tryInvoke" - like do nothing

(I think we should move alternative solutions for tryInvoke into its own section)

Reply:
For Ember assert, I think of this case as: the function argument is **required** and not optional. Although its good practice to make all function arguments as required, I wanted to use alternative examples that were truly optional for the Ember community.

Reply:
We can definitely move it in its own section. Does `Alternative Solutions` as a heading work?

Reply:
Yeah totally!